src/sentry/debug_files/artifact_bundle_indexing.py

```diff
@@ -216,21 +235,105 @@ def mark_bundle_for_flat_file_indexing(
 
 @sentry_sdk.tracing.trace
 def remove_artifact_bundle_from_indexes(artifact_bundle: ArtifactBundle):
+    redis_client = get_redis_cluster_for_artifact_bundles()
+
     flat_file_indexes = ArtifactBundleFlatFileIndex.objects.filter(
         flatfileindexstate__artifact_bundle=artifact_bundle
     )
     for idx in flat_file_indexes:
-        identifier = FlatFileIdentifier(
-            project_id=idx.project_id, release=idx.release_name, dist=idx.dist_name
-        )
+        identifier = FlatFileIdentifier.from_index(idx)
 
         was_removed = update_artifact_bundle_index(
             identifier, bundles_to_remove=[artifact_bundle.id]
         )
         if not was_removed:
             metrics.incr("artifact_bundle_flat_file_indexing.removal.would_block")
-            # TODO: mark this for async removal
-            pass
+            redis_client.sadd(get_deletion_key(idx.id), artifact_bundle.id)
+            redis_client.sadd(get_deletion_key(), idx.id)
+
+
+@sentry_sdk.tracing.trace
+def backfill_artifact_index_updates() -> bool:
+    redis_client = get_redis_cluster_for_artifact_bundles()
+
+    indexes_needing_update = list(
```

Comment:
Should we maybe use a date to check for `NOT_INDEXED` since we would like to avoid concurrent updates. Maybe like, if it has not been indexed for 10 mins, remove index it, otherwise don't do it in this run.

Reply:
We might end up with some overlap in situations in which we have the job overlapped with uploads but it's fine if we do more work.